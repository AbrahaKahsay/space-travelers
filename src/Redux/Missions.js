// Actions
const SPACE_MISSIONS = 'SPACE_MISSIONS';
const MISSIONS_RESERVED = 'MISSIONS_RESERVED';
const MISSIONS_CANCELLED = 'MISSIONS_CANCELLED';

const url = 'https://api.spacexdata.com/v3/missions/';

export const fetchMissions = () => async (dispatch) => {
    const data = await fetch(url);
    const response = await data.json();
    dispatch(
        getMissions(
            response.map((mission) => ({
                id: mission.mission_id,
                name: mission.mission_name,
                description: mission.description,
                reserved: false,
            })),
        ),
    );
};

// Action creators
export const reserveMission = (id) => ({
    type: MISSIONS_RESERVED,
    id,
});

export const cancelMission = (id) => ({
    type: MISSIONS_CANCELLED,
    id,
});

export const getMissions = (missions) => ({
    type: SPACE_MISSIONS,
    missions,
});

const initialState = [];
let resState;
// reducer
const missionsReducer = (state = initialState, action) => {
    switch (action.type) {
        case SPACE_MISSIONS:
            return action.missions;

        case MISSIONS_RESERVED:
            resState = state.map((item) => {
                if (item.id !== action.id) return item;
                return { ...item, reserved: true };
            });
            return resState;
        case MISSIONS_CANCELLED:
            resState = state.map((item) => {
                if (item.id !== action.id) return item;
                return { ...item, reserved: false };
            });
            return resState;

        default:
            return state;
    }
};

export default missionsReducer;