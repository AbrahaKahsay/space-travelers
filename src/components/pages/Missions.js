import Mission from '../Mission';
import '../styles/Mission.css';

const Missions = () => (
  <div className="mission">
    <Mission />
  </div>
);

export default Missions;
