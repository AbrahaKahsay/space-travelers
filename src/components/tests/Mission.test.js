import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import Missions from '../Mission/Mission';
import store from '../../Redux/configureStore';

describe('Jest Snapshot testing suite', () => {
  it('Matches DOM Snapshot', () => {
    const { container } = render(
      <MemoryRouter>
        <Provider store={store}>
          <Missions />
        </Provider>
      </MemoryRouter>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});