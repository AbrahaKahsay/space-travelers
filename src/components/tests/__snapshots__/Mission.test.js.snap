// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jest Snapshot testing suite Matches DOM Snapshot 1`] = `
<table
  cellspacing="0"
  class="table"
>
  <thead
    class="table-head"
  >
    <tr>
      <th>
        Mission
      </th>
      <th>
        Description
      </th>
      <th>
        Status
      </th>
      <th>
        <div />
      </th>
    </tr>
  </thead>
  <tbody
    class="table-body"
  />
</table>
`;
